californium-core/src/main/java/org/eclipse/californium/core/network/InMemoryMessageExchangeStore.java

```diff
@@ -61,13 +61,12 @@
 	private final ConcurrentMap<KeyToken, Exchange> exchangesByToken = new ConcurrentHashMap<>();
 
 	private final NetworkConfig config;
+	private final TokenProvider tokenProvider;
 	private boolean running = false;
-	private Deduplicator deduplicator;
+	private volatile Deduplicator deduplicator;
```

Comment:
From my view, the intended usage of this class according threads seems to be not clear, or at least not documented.
The `deduplicator` write-access is synchronized and done before or with `start`.
The `deduplicator` read-access is not synchronized (`find `or `findPrevious`). I don't see a comment, that this methods are only allowed to be called from threads which are created after the `start` and even if, it would be hard to ensure that.
So in my opinion the easiest way to ensure, that other threads read the right value is `volatile`.
A other approach would be to introduce a "synchronized getter".